*: Allow exposing multiple label-sets

brancz opened the pull request:
## Changes 

This PR adds that the Info gRPC service now exposes potentially multiple label-sets. For all store types except `proxy` this doesn't change anything for now, it just adds the currently single label-set as the only entry in the list of label-sets. The `proxy` type, from now exposes it's stores' labels merged with its configured selector.

I'd like to ask that the `proxy.go` changes are particularly carefully reviewed.

## Verification

Adapted all tests and see them passing.

@squat @bwplotka @GiedriusS @povilasv @domgreen

povilasv commented on pkg/store/storepb/rpc.proto:
```diff
@@ -42,9 +42,14 @@ enum StoreType {
 
 message InfoResponse {
   repeated Label labels = 1 [(gogoproto.nullable) = false];
```
This one should be marked as deprecated no?

povilasv replied:
something like `reserved 0;`

bwplotka replied:
:+1:

brancz replied:
Commented as deprecated.

povilasv commented on pkg/store/proxy.go:
```diff
@@ -104,15 +123,69 @@ func (s *ProxyStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb
+	}
+
+	g, gctx := errgroup.WithContext(ctx)
+	respSender, respRecv, closeFn := newInfoRespCh(gctx, 10)
```
I'm a bit worried about this part, whether it will actually work with all the time outs and partial response strategy.

So IMO more tests is needed.

Also maybe this could be simplified using mutex? Channels are a bit hard to deal with :/

brancz replied:
I don't think there is a possibility for tolerating partial responses. We essentially need a consistent system for metadata like this, meaning either all succeed or none, and metadata stays unchanged until the next successful request.

Adding more tests is definitely a good idea, on that.

bwplotka replied:
Hm.. Why we are doing this here? (: 

Querier already maintain healthy nodes in storeset. It does it by periodic `Info` calls. This always include labels as well. I think we can rely on this asyncronous Info calls here as well instead of duplicated, sync forwarding. What do you think? (:

I think some lag is fine with benefits of reduced number of requests + simplified code here.

brancz replied:
That's fair, will adapt.

brancz replied:
Removed all the async code here and just assembled the label-sets with the cached store's label-sets.

bwplotka commented on pkg/store/storepb/rpc.proto:
```diff
@@ -42,9 +42,14 @@ enum StoreType {
-  int64 min_time        = 2;
-  int64 max_time        = 3;
-  StoreType storeType   = 4;
+  int64 min_time = 2;
```
Our manual pretty indent is gone ;p

not a blocker though.

brancz replied:
I tried, but it started looking weirder and weirder. Let me know if it bothers you :)

bwplotka replied:
Nah, I think your time is way more precious to waste it on  manual linting (:

bwplotka commented on pkg/store/storepb/rpc.proto:
```diff
@@ -42,9 +42,14 @@ enum StoreType {
 
 message InfoResponse {
   repeated Label labels = 1 [(gogoproto.nullable) = false];
```
Let's comment as Deprecated

brancz replied:
Done

bwplotka commented on test/e2e/rule_test.go:
```diff
@@ -241,6 +241,16 @@ func (a *failingStoreAPI) Info(context.Context, *storepb.InfoRequest) (*storepb.
 		},
+		LabelSets: []storepb.LabelSet{
+			{
+				Labels: []storepb.Label{
```
Let's make test more ambitious and have multiple set here?

brancz replied:
Done

bwplotka commented on pkg/query/storeset.go:
```diff
@@ -238,10 +239,12 @@ func (s *StoreSet) Update(ctx context.Context) {
 			store.close()
 			s.updateStoreStatus(store, errors.New(droppingStoreMessage))
-			level.Warn(s.logger).Log("msg", droppingStoreMessage, "address", addr)
+			level.Warn(s.logger).Log("msg", droppingStoreMessage, "address", addr, "externalLabels", externalLabels, "duplicate-amount", storesWithExternalLabels)
```
```suggestion
			level.Warn(s.logger).Log("msg", droppingStoreMessage, "address", addr, "extLset", externalLabels, "duplicates", storesWithExternalLabels)
```
I feel like short names are fine + let's be consistent on camelCase, snakeCase

brancz replied:
Done.

bwplotka commented on pkg/query/storeset.go:
```diff
@@ -238,10 +239,12 @@ func (s *StoreSet) Update(ctx context.Context) {
-		if len(store.Labels()) > 0 && externalLabelStores[externalLabelsFromStore(store)] != 1 {
+		externalLabels := externalLabelsFromStore(store)
+		storesWithExternalLabels := externalLabelStores[externalLabels]
+		if len(store.LabelSets()) > 0 && len(store.LabelSets()[0].Labels) > 0 && storesWithExternalLabels != 1 {
```
`len(store.LabelSets()[0].Labels) > 0 ` -> hm, what if `1` will have 0 elements? (:

I would skip this one. Let's make this natural, so we don't expect external labels to be empty? if empty the labelset is empty? What do you think?

brancz replied:
Yes I was thinking something similar, and with the other invariant in tsdb.go and prometheus.go this works out nicely.

brancz replied:
Done. It's much nicer this way.

bwplotka commented on pkg/store/tsdb.go:
```diff
@@ -58,6 +58,14 @@ func (s *TSDBStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb.
+	// them here for migration/compatibility purposes.
+	res.LabelSets = []storepb.LabelSet{
+		{
+			Labels: res.Labels,
```
I would aim for invariant that if `res.Labels` are empty then `res.LabelSets` is `[]` - it will simplify interface a bit. WDYT?

brancz replied:
Yes agreed.

brancz replied:
Done.

bwplotka commented on pkg/query/storeset.go:
```diff
@@ -331,16 +334,26 @@ func (s *StoreSet) getHealthyStores(ctx context.Context) map[string]*storeRef {
+	}
+	sort.Strings(tsdbLabelSetStrings)
+
+	s := ""
```
can we replace this with `return strings.Join(tsdbLabelSetStrings, ", )` or some other delimiter?

bwplotka replied:
or just "" delimiter to have exactly same thing as you have here.

brancz replied:
:man_facepalming: you're absolutely right

brancz replied:
Done.

bwplotka commented on pkg/store/prometheus.go:
```diff
@@ -98,6 +98,14 @@ func (p *PrometheusStore) Info(ctx context.Context, r *storepb.InfoRequest) (*st
+
+	// Until we deprecate the single labels in the reply, we just duplicate
+	// them here for migration/compatibility purposes.
+	res.LabelSets = []storepb.LabelSet{
```
ditto as tsdb.go

brancz replied:
Done.

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -104,15 +123,69 @@ func (s *ProxyStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb
 	res.MaxTime = MaxTime
```
I know it is not yours, but can we fix above code? It looks weird that we loop over `range store` then check `len(stores)` 2 times. 

Also `MaxTime -> maxTime`, `MinTime ->  minTime`

brancz replied:
Moved all `len(stores) == 0` to an early return before anything else.

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -104,15 +123,69 @@ func (s *ProxyStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb
 	res.MaxTime = MaxTime
 	res.MinTime = MinTime
 
-	for _, l := range s.selectorLabels {
```
We need to remove selector flag as well then + CHANGELOG

brancz replied:
Actually didn't remove this. If specified label-sets are merged with the selector labels specified.

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -74,6 +74,25 @@ func NewProxyStore(
+	ch  chan<- *storepb.InfoResponse
+}
+
+func newInfoRespCh(ctx context.Context, buffer int) (*ctxInfoRespSender, <-chan *storepb.InfoResponse, func()) {
```
might be not needed as commented above

brancz replied:
Removed all of this async code.

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -193,7 +266,7 @@ func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesSe
 			sc, err := st.Series(seriesCtx, r)
 			if err != nil {
-				storeID := fmt.Sprintf("%v", storepb.LabelsToString(st.Labels()))
+				storeID := storepb.LabelSetsToString(st.LabelSets())
```
:+1:

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -388,18 +461,20 @@ func storeMatches(s Client, mint, maxt int64, matchers ...storepb.LabelMatcher)
-				continue
-			}
+		for _, ls := range s.LabelSets() {
+			for _, l := range ls.Labels {
```
hm, something is wrong here - or if right we need to comment it better.

Essentially this means that matching here treat set with `AND` operator and I think we need `OR` rather, right?

as:
`[a=1, b=1][a=1, b=2]` labelset should be *matched* for `a=1, b=2` matcher, but it will be fitered out by `if !m.Matches(l.Value) {`

brancz replied:
Yes you're right.

brancz replied:
Refactored this a bunch, I think this is much clearer now.

bwplotka commented on pkg/store/storepb/custom.go:
```diff
@@ -177,3 +178,11 @@ func LabelsToString(lset []Label) string {
+
+func LabelSetsToString(lsets []LabelSet) string {
+	s := ""
+	for _, ls := range lsets {
```
I think convention in Thanos is to use `strings.Join` but here we might use `bytes.Buffer.WriteString` I guess? I think we don't care about performance though as it's only for printing. Not a blocker, just `+` for strings feels wrong in Golang although it's heavily optimized nowadays.

brancz replied:
Replaced with `strings.Join` based approached.

brancz commented on pkg/query/storeset.go:
```diff
@@ -215,6 +215,7 @@ func (s *StoreSet) Update(ctx context.Context) {
 	for _, st := range healthyStores {
 		externalLabelStores[externalLabelsFromStore(st)]++
 	}
+	level.Debug(s.logger).Log("externalLabelStores", fmt.Sprintf("%#+v", externalLabelStores))
```
Done.

brancz commented:
I believe I have addressed all comments and rebased the branch to resolve conflicts with master. Let me know what you think :slightly_smiling_face: . Overall this is in a much better shape than the first time around I think! I'm feeling much more comfortable with it now. Thanks for the reviews!

bwplotka commented on CHANGELOG.md:
```diff
@@ -19,6 +19,8 @@ We use *breaking* word for marking changes that are not backward compatible (rel
 
 - [#1248](https://github.com/improbable-eng/thanos/pull/1248) Add a web UI to show the state of remote storage.
 
+- [#1284](https://github.com/improbable-eng/thanos/pull/1284) Add support for multiple label-sets in Info gRPC service. This deprecates the single `Labels` slice of the `InfoResponse`.
```
Let's mention maybe what's the plan on deprecation. It looks like breaking change, while it is not. Worth to add details on that otherwise users will be like `Ok, but then all the components has to be upgraded to 0.6.0 if I use 0.6.0 Querier`?

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -81,38 +81,73 @@ func (s *ProxyStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb
 	stores := s.stores()
+
+	if len(stores) == 0 {
+		res.MaxTime = math.MaxInt64
```
I am not sure about this. I know there is not regression to previous version (same logic), but does this make sense? ;p
Let's leave the comment maybe? `// Edge case: we have all of the data if there are no stores.` and add some `TODO`. 

Essentially not sure if we no store we should do either
* max time = MaxInt64 and min time = maxInt64 as nothing is there.
or
* max time = MaxInt64 and min Time MinInt64 if we want to have `whole range` ;p 

But we should not spend time on this in this PR I guess.

brancz replied:
Would an issue be more appropriate than a TODO in code maybe?

brancz replied:
I'm adding the "Edge case:" comment back and will open an issue to decide this as a follow up.

bwplotka replied:
Thanks, makes sense

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -81,38 +81,73 @@ func (s *ProxyStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb
+	res.MaxTime = maxTime
+	res.MinTime = minTime
 
 	for _, l := range s.selectorLabels {
```
I assume we are not removing `selector Labels` for now?

brancz replied:
It somewhat depends on our decision around the changelog. If we decide all queriers must be >=v0.6.0 if any are, then this would be safe to remove.

bwplotka replied:
:+1:  Let's leave for now

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -81,38 +81,73 @@ func (s *ProxyStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb
+
+	res.LabelSets = make([]storepb.LabelSet, 0, len(labelSets))
+	for _, v := range labelSets {
+		res.LabelSets = append(res.LabelSets, storepb.LabelSet{Labels: v})
```
Worth to mention on interface that sets are not sorted?

brancz replied:
Added to the proto message and generated structs.

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -81,38 +81,73 @@ func (s *ProxyStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb
 }
 
+// mergeLabels merges label-sets a and b with label-set b having precedence.
+func mergeLabels(labelSet []storepb.Label, selector labels.Labels) []storepb.Label {
```
I guess `a []storepb.Label, b labels.Labels)` would makes sense then more by looking on comment?

bwplotka commented on pkg/store/proxy.go:
```diff
@@ -381,14 +416,39 @@ func (s *streamSeriesSet) Err() error {
+	return labelSetsMatch(s.LabelSets(), matchers)
+}
+
+// labelSetsMatch returns false if each label-set negative matches against the
```
I would change this to (find it more clear for me), but up to you (: 
```suggestion
// labelSetsMatch returns false if all label-set does not match the
```

bwplotka commented on pkg/query/storeset.go:
```diff
@@ -32,14 +33,14 @@ type StoreSpec interface {
 	// NOTE: It is implementation responsibility to retry until context timeout, but a caller responsibility to manage
 	// given store connection.
-	Metadata(ctx context.Context, client storepb.StoreClient) (labels []storepb.Label, mint int64, maxt int64, err error)
+	Metadata(ctx context.Context, client storepb.StoreClient) (labelSets []storepb.LabelSet, labels []storepb.Label, mint int64, maxt int64, err error)
```
I don't really get why not just replacing with labelSets.

We can hide the compatibility logic internally, right?

bwplotka commented on pkg/query/storeset.go:
```diff
@@ -162,27 +163,31 @@ type storeRef struct {
-	s.mtx.RLock()
-	defer s.mtx.RUnlock()
+func (s *storeRef) Update(labelSets []storepb.LabelSet, labels []storepb.Label, minTime int64, maxTime int64) {
+	if len(labelSets) == 0 && len(labels) > 0 {
```
I would hide it behind Metadata interface.

bwplotka commented:
Awesome! :ship: it.

Thanks :+1:
